firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtil.java

```diff
@@ -0,0 +1,213 @@
+      return getDownloadingCustomModelDetails(modelName);
+    }
+
+    String filePath =
+        getSharedPreferences()
+            .getString(String.format(LOCAL_MODEL_FILE_PATH_PATTERN, persistenceKey, modelName), "");
+
+    long fileSize =
+        getSharedPreferences()
+            .getLong(String.format(LOCAL_MODEL_FILE_SIZE_PATTERN, persistenceKey, modelName), 0);
+
+    // if no-zero - local model is present and new model being downloaded
+    long id =
+        getSharedPreferences()
+            .getLong(String.format(DOWNLOADING_MODEL_ID_PATTERN, persistenceKey, modelName), 0);
+
+    return new CustomModel(modelName, id, fileSize, modelHash, filePath);
+  }
+
+  /**
+   * Returns the Custom Model details associated with this version of this model currently being
+   * downloaded. If no download is in progress return null. Contains no information about local
+   * model, only download status.
+   *
+   * @param modelName name of the model
+   * @return Download version of CustomModel
+   */
+  @Nullable
+  public synchronized CustomModel getDownloadingCustomModelDetails(@NonNull String modelName) {
+    String modelHash =
+        getSharedPreferences()
+            .getString(
+                String.format(DOWNLOADING_MODEL_HASH_PATTERN, persistenceKey, modelName), null);
+
+    if (modelHash == null || modelHash.isEmpty()) {
+      // no model hash means no download in progress
+      return null;
+    }
+
+    long fileSize =
+        getSharedPreferences()
+            .getLong(String.format(DOWNLOADING_MODEL_SIZE_PATTERN, persistenceKey, modelName), 0);
+
+    long id =
+        getSharedPreferences()
+            .getLong(String.format(DOWNLOADING_MODEL_ID_PATTERN, persistenceKey, modelName), 0);
+
+    return new CustomModel(modelName, id, fileSize, modelHash);
+  }
+
+  /**
+   * The information about the new custom model download that need to be stored.
+   *
+   * @param customModel custom model details to be stored.
+   */
+  public synchronized void setDownloadingCustomModelDetails(@NonNull CustomModel customModel) {
+    String modelName = customModel.getName();
+    String modelHash = customModel.getModelHash();
+    long downloadId = customModel.getDownloadId();
+    long modelSize = customModel.getSize();
+    getSharedPreferences()
+        .edit()
+        .putString(
+            String.format(DOWNLOADING_MODEL_HASH_PATTERN, persistenceKey, modelName), modelHash)
+        .putLong(
+            String.format(DOWNLOADING_MODEL_SIZE_PATTERN, persistenceKey, modelName), modelSize)
+        .putLong(String.format(DOWNLOADING_MODEL_ID_PATTERN, persistenceKey, modelName), downloadId)
+        // The following assumes the download will finish before the system reboots.
+        // If not, the download duration won't be correct, which isn't critical.
+        .putLong(
+            String.format(DOWNLOAD_BEGIN_TIME_MS_PATTERN, persistenceKey, modelName),
+            SystemClock.elapsedRealtime())
+        .apply();
```

Comment:
`apply()` only alters the in-memory contents of the SharedPreferences, and eventually writes to disk, which may fail silently, unlike `commit()` that's synchronous. Just want to double check that this is an acceptable situation in this case.

Reply:
changed to commit